// idyll-components/dist/cjs/action.js
'use strict';

var _extends = require('@babel/runtime/helpers/extends');
var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

class Action extends React__default["default"].PureComponent {
  render() {
    const {
      idyll,
      hasError,
      updateProps,
      ...props
    } = this.props;
    return /*#__PURE__*/React__default["default"].createElement("span", _extends__default["default"]({}, props, {
      onClick: this.props.onClick ? this.props.onClick : e => e.stopPropagation(),
      className: 'idyll-action'
    }), this.props.children);
  }

}

Action._idyll = {
  name: 'Action',
  tagType: 'open',
  displayType: 'inline',
  children: ['Click me to trigger an action'],
  props: [{
    name: 'onClick',
    type: 'event',
    example: '`alert("You clicked the text.")`',
    description: 'An event that is triggered when a reader clicks.'
  }, {
    name: 'onMouseEnter',
    type: 'event',
    description: 'An event that is triggered when a reader mouses over the element, e.g. `x = true`.'
  }, {
    name: 'onMouseLeave',
    type: 'event',
    description: 'An event that is triggered when a reader removes their mouse from the element, e.g. `x = false`.'
  }]
};

module.exports = Action;
