// idyll-components/dist/cjs/boolean.js
'use strict';

var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

class Boolean extends React__default["default"].PureComponent {
  constructor(props) {
    super(props);
  }

  toggleCheckbox() {
    this.props.updateProps({
      value: !this.props.value
    });
  }

  render() {
    const {
      value,
      className,
      style
    } = this.props;
    return /*#__PURE__*/React__default["default"].createElement("input", {
      type: "checkbox",
      onChange: this.toggleCheckbox.bind(this),
      onClick: this.props.onClick || (e => e.stopPropagation()),
      checked: value,
      className: `idyll-checkbox ${className ? className : ''}`.trim(),
      style: style
    });
  }

}

Boolean.defaultProps = {
  value: false
};
Boolean._idyll = {
  name: 'Boolean',
  tagType: 'closed',
  props: [{
    name: 'value',
    type: 'boolean',
    example: 'x',
    description: 'A value for the checkbox. If this value is truthy, the checkbox will be shown.'
  }]
};

module.exports = Boolean;
