// idyll-components/dist/cjs/button.js
'use strict';

var _extends = require('@babel/runtime/helpers/extends');
var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

class Button extends React__default["default"].PureComponent {
  render() {
    const {
      onClick,
      idyll,
      hasError,
      updateProps,
      ...props
    } = this.props;
    return /*#__PURE__*/React__default["default"].createElement("button", _extends__default["default"]({}, props, {
      onClick: onClick.bind(this)
    }), this.props.children);
  }

}

Button.defaultProps = {
  onClick: function () {}
};
Button._idyll = {
  name: 'Button',
  tagType: 'open',
  children: ['Click Me.'],
  props: [{
    name: 'onClick',
    type: 'event',
    example: '`x += 1`',
    description: 'An event that is fired when the user clicks the button.'
  }]
};

module.exports = Button;
