// idyll-components/dist/cjs/code-highlight.js
'use strict';

var React = require('react');
var SyntaxHighlighter = require('react-syntax-highlighter/dist/light');
var style = require('react-syntax-highlighter/dist/styles/github');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var SyntaxHighlighter__default = /*#__PURE__*/_interopDefaultLegacy(SyntaxHighlighter);
var style__default = /*#__PURE__*/_interopDefaultLegacy(style);

class CodeHighlight extends React__default["default"].PureComponent {
  render() {
    return /*#__PURE__*/React__default["default"].createElement(SyntaxHighlighter__default["default"], {
      style: style__default["default"],
      language: this.props.language
    }, this.props.children.length ? this.props.children[0] : '');
  }

}

CodeHighlight.defaultProps = {
  children: []
};
CodeHighlight._idyll = {
  name: "CodeHighlight",
  tagType: "open",
  children: ['var x = 1;'],
  props: [{
    name: "language",
    type: "string",
    example: "javascript"
  }]
};

module.exports = CodeHighlight;
