// idyll-components/dist/cjs/float.js
'use strict';

var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

class Float extends React__default["default"].PureComponent {
  render() {
    return /*#__PURE__*/React__default["default"].createElement("div", {
      className: `float ${this.props.position}`,
      style: {
        float: this.props.position,
        width: this.props.width || '50%',
        ...this.props.style
      }
    }, this.props.children);
  }

}

Float._idyll = {
  name: 'Float',
  tagType: 'open',
  children: [`Content placed here will take up half of the text-column width. Use the properties to specify left or right.`],
  props: [{
    name: 'position',
    type: 'string',
    example: '"left"',
    description: 'the float position: left or right.'
  }, {
    name: 'width',
    type: 'string',
    example: '"50%"',
    defaultValue: '"50%"',
    description: 'the width of the component, specified in pixels or percentage.'
  }]
};

module.exports = Float;
