// idyll-components/dist/cjs/h2.js
'use strict';

var _extends = require('@babel/runtime/helpers/extends');
var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

const generateId = (headerText = '') => {
  return headerText.toString().trim().replace(/\s+/g, '-').toLowerCase();
};

const GenerateHeaders = props => {
  const {
    idyll,
    hasError,
    updateProps,
    size,
    children = [],
    ...attributeProps
  } = props;
  const headerText = children.join('');
  const HeaderTag = `h${size}`;

  if (!attributeProps.id) {
    attributeProps.id = generateId(headerText);
  }

  return /*#__PURE__*/React__default["default"].createElement(HeaderTag, attributeProps, children);
};

class H2 extends React__default["default"].PureComponent {
  render() {
    return /*#__PURE__*/React__default["default"].createElement(GenerateHeaders, _extends__default["default"]({
      size: "2"
    }, this.props));
  }

}

H2._idyll = {
  name: 'H2',
  tagType: 'open',
  children: ['My Header Size 2']
};

module.exports = H2;
