// idyll-components/dist/cjs/select.js
'use strict';

var _extends = require('@babel/runtime/helpers/extends');
var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

require('react-dom');

class Select extends React__default["default"].PureComponent {
  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
  }

  onChange(e) {
    this.props.updateProps({
      value: e.target.value
    });
  }

  render() {
    const {
      idyll,
      hasError,
      updateProps,
      ...props
    } = this.props;
    return /*#__PURE__*/React__default["default"].createElement("select", _extends__default["default"]({
      onChange: this.onChange
    }, props, {
      onClick: this.props.onClick || (e => e.stopPropagation())
    }), this.props.options.map(d => {
      if (typeof d === 'string') {
        return /*#__PURE__*/React__default["default"].createElement("option", {
          key: d,
          value: d
        }, d);
      }

      return /*#__PURE__*/React__default["default"].createElement("option", {
        key: d.label || d.value,
        value: d.value
      }, d.label || d.value);
    }));
  }

}

Select.defaultProps = {
  options: []
};
Select._idyll = {
  name: 'Select',
  tagType: 'closed',
  props: [{
    name: 'value',
    type: 'string',
    example: 'x',
    description: 'The currently selected value.'
  }, {
    name: 'options',
    type: 'array',
    example: '`["option1", "option2"]`',
    description: 'An array representing the different options. Can be an array of strings like `["val1", "val2"]` or an array of objects `[{ value: "val1", label: "Value 1" }, { value: "val2", label: "Value 2" }]`.'
  }]
};

module.exports = Select;
