// idyll-components/dist/cjs/step.js
'use strict';

var _extends = require('@babel/runtime/helpers/extends');
var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

let idx = 0;

class Step extends React__default["default"].Component {
  componentDidMount() {
    let _idx = this.props.stepIndex === undefined ? idx++ : this.props.stepIndex;

    this.props.registerStep && this.props.registerStep(_idx, this.props.state, (this.props.onEnter || (() => {})).bind(this));
  }

  render() {
    const {
      idyll,
      updateProps,
      hasError,
      registerStep,
      onEnter,
      state,
      className,
      stepIndex,
      ...props
    } = this.props;
    return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
      ref: ref => this.ref = ref,
      className: `idyll-step ${className || ''}`
    }, props));
  }

}

Step._idyll = {
  name: 'Step',
  tagType: 'open'
};

module.exports = Step;
